crates/oci/src/client.rs

```diff
@@ -246,43 +273,59 @@ impl Client {
         &mut self,
         locked: &mut LockedApp,
         assembly_mode: AssemblyMode,
+        compose_mode: ComposeMode,
     ) -> Result<Vec<ImageLayer>> {
         let mut layers = Vec::new();
         let mut components = Vec::new();
         for mut c in locked.clone().components {
-            // Add the wasm module for the component as layers.
-            let source = c
-                .clone()
-                .source
-                .content
-                .source
-                .context("component loaded from disk should contain a file source")?;
+            match compose_mode {
```

Comment:
This stanza is very long and deeply nested, and by the time I got to the end of it I had kind of lost track of where I was or how it fitted into the flow.  (E.g. does line 328 end the for loop or only the match?  I need to scroll a whole screen up and squint at braces to tell.)  It would be really good to see if we could break the loop body out into two or three helper functions e.g. 

```
for ... {
  match compose_mode {
    All => composify(),
    Skip => dont_composify(),
  };
  assemble_layers_for_one();
}
```

kind of thing.  Or whatever you feel the appropriate cohesive units are.